### PR TITLE
Improve Methods in NullOpsDecorator

### PR DESCRIPTION
**Pull request by noti0na1**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala [hunk 1]
```diff
@@ -8,16 +8,35 @@ import dotty.tools.dotc.core.Types.{AndType, ClassInfo, ConstantType, OrType, Ty
 object NullOpsDecorator {
 
   implicit class NullOps(val self: Type) {
+    def findAndStripNull(implicit ctx: Context): (Type, Boolean, Boolean) = {
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala]: I thought we're changing this back to `normNullableUnion` and returning just a `Type`?

Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala]: I am still working on this PR

**Comment by noti0na1** [comment]
@abeln could you have look at updates?

File: compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala [hunk 2]
```diff
@@ -8,18 +8,10 @@ import dotty.tools.dotc.core.Types.{AndType, ClassInfo, ConstantType, OrType, Ty
 object NullOpsDecorator {
 
   implicit class NullOps(val self: Type) {
+
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala]: nit: remove please (anything that minimizes the diff's size is good :-) )

File: compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala [hunk 3]
```diff
@@ -29,20 +21,64 @@ object NullOpsDecorator {
       self.isDirectRef(defn.JavaNullAlias)
     }
 
+    /** Normalizes unions so that all `Null`s (or aliases to `Null`) appear to the right of
+     *  all other types.
+     *  e.g. `Null | (T1 | Null) | T2` => `T1 | T2 | Null`
+     *  e.g. `JavaNull | (T1 | Null) | Null` => `T1 | Null | JavaNull`
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala]: I think we can simplify this to `T1 | JavaNull`. JavaNull should only appear on Java methods, in which case Null won't appear (because there's no Null type in Java).

File: compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala [hunk 4]
```diff
@@ -29,20 +21,64 @@ object NullOpsDecorator {
...
+     *
+     *  1. If self is not a union type, the result is not a union type.
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala]: Let's rephrase the method contract as follows.

```
Let `self` denote the current type:
  1. If `self` is not a union, then the result is `self`.
  2. If `self` is a union then
    2.1 If `self` does not contain `Null` as part of the union, then the result is `self`.
    2.2 If `self` contains `Null` (resp `JavaNull`) as part of the union, let `self2` denote the same type as `self`,
          but where all instances of `Null` (`JavaNull`) in the union have been removed. Then the result is `self2 | Null` (`self2 | JavaNull`).
```

File: compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala [hunk 5]
```diff
@@ -29,20 +21,64 @@ object NullOpsDecorator {
...
+          if (tp.isNullType) {
+            if (tp.isJavaNullType) hasJavaNull = true
+            else hasNull = true
+          }
+          tp
+      }
+      val tp = strip(self)
+      if (!isUnion) self
+      else {
+        val tp1 = if (hasNull) OrType(tp, defn.NullType) else tp
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala]: This will need to change to only return the `JavaNull` if both Null and JavaNull are present.

File: compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala [hunk 6]
```diff
@@ -59,65 +95,33 @@ object NullOpsDecorator {
...
-      val (tp, diff) = strip(self.widenDealias.normNullableUnion, changed = false)
-      if (diff) tp else self
-    }
-
-    /** Syntactically strips the nullability from this type. If the normalized form (as per `normNullableUnion`)
-     *  of this type is `T1 | ... | Tk | ... | Tn`, and all types in the range `Tk ... Tn` are references to `Null`,
-     *  then return `T1 | ... | Tk-1`.
+    /** Syntactically strips the nullability from this type.
+     *  If the normalized form (as per `normNullableUnion`) of this type is `T1 | ... | Tn-1 | Tn`,
+     *  and `Tn-1` `Tn` are references to `Null`, then return `T1 | ... | Tn-2`.
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala]: Will need to update the comment to reflect the fact that only the last component of the union will be Null or JavaNull.

File: compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala [hunk 7]
```diff
@@ -59,65 +95,33 @@ object NullOpsDecorator {
...
      *  If this type isn't (syntactically) nullable, then returns the type unchanged.
      */
     def stripNull(implicit ctx: Context): Type = {
       assert(ctx.explicitNulls)
-      stripNullImpl(onlyJavaNull = false)
+      self.widenDealias.normNullableUnion match {
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala:101]: Can simplify as above.

File: compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala [hunk 8]
```diff
@@ -129,12 +133,14 @@ object NullOpsDecorator {
...
       object RemoveNulls extends TypeMap {
-        override def apply(tp: Type): Type = tp match {
-          case tp: OrType if tp.isJavaNullableUnion =>
-            diff = true
-            mapOver(tp.stripJavaNull)
-          case _ => mapOver(tp)
-        }
+        override def apply(tp: Type): Type =
+          tp.normNullableUnion match {
+            // It it is JavaNullable Union
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala]: nit: the comment is redundant and can be removed

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 9]
```diff
@@ -1068,18 +1068,28 @@ object Types {
...
-        } else {
+        lazy val defaultp =
           ctx.typeComparer.lub(tp1.widenUnion, tp2.widenUnion, canConstrain = true) match {
             case union: OrType => union.join
             case res => res
           }
+        if (ctx.explicitNulls) {
+          // Don't widen `T|Null`, since otherwise we wouldn't be able to infer nullable unions.
+          tp.normNullableUnion match {
+            // If it is a nullable union
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: nit: redundant comment

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 10]
```diff
@@ -1068,18 +1068,28 @@ object Types {
...
+            case OrType(lhs, rhs) if rhs.isNullType =>
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: Can simplify since only the last component can be null.

File: tests/explicit-nulls/neg/nullnull.scala [hunk 11]
```diff
@@ -0,0 +1,9 @@
+// Test that Null | Null will not cause crash during typing.
```
Comment (abeln) [inline tests/explicit-nulls/neg/nullnull.scala]: Could you please add a comment explaining what happens to `Null | Null`?

File: tests/explicit-nulls/neg/nullnull.scala [hunk 12]
```diff
@@ -0,0 +1,9 @@
...
+
+class Foo {
+  def foo1: Unit = {
+    val x: Null | Null | Null = ???
```
Comment (abeln) [inline tests/explicit-nulls/neg/nullnull.scala:8]: Could you please add another test where `x` has type "String | JavaNull | Null", where we remember that x was a `JavaNull`?

**Review by abeln — CHANGES_REQUESTED** [review]
Thanks for looking into this. I have a few comments.

**Comment by noti0na1** [comment]
@abeln  Done

File: compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala [hunk 13]
```diff
@@ -29,20 +20,62 @@ object NullOpsDecorator {
...
+    }
+
     /** Is self (after widening and dealiasing) a type of the form `T | JavaNull`? */
     def isJavaNullableUnion(implicit ctx: Context): Boolean = {
       assert(ctx.explicitNulls)
-      def peelOff(tpe: Type): Boolean = tpe match {
-        case OrType(left, right) =>
-          right.isJavaNullType || right.isNullType && peelOff(left)
+      self.widenDealias.normNullableUnion match {
+        // If it a `JavaNull` Union, the most right type must be a `JavaNull`.
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala]: nit: remove the redundant comment

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 14]
```diff
@@ -1068,26 +1068,36 @@ object Types {
...
      */
-    def widenUnion(implicit ctx: Context): Type = widen match {
-      case tp @ OrType(tp1, tp2) =>
-        if (ctx.explicitNulls && tp.isNullableUnion) {
-          // Don't widen `T|Null`, since otherwise we wouldn't be able to infer nullable unions.
-          val OrType(leftTpe, nullTpe) = tp.normNullableUnion
-          OrType(leftTpe.widenUnion, nullTpe)
-        } else {
-          ctx.typeComparer.lub(tp1.widenUnion, tp2.widenUnion, canConstrain = true) match {
+    def widenUnion(implicit ctx: Context): Type = {
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/Types.scala:1071]: Let's try and get rid of the call to `normNullableUnion` as we discussed in person.

**Review by abeln — CHANGES_REQUESTED** [review]

**Comment by noti0na1** [comment]
@abeln `widenUnion` updated

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 15]
```diff
@@ -1068,26 +1068,50 @@ object Types {
...
-        if (ctx.explicitNulls && tp.isNullableUnion) {
-          // Don't widen `T|Null`, since otherwise we wouldn't be able to infer nullable unions.
-          val OrType(leftTpe, nullTpe) = tp.normNullableUnion
-          OrType(leftTpe.widenUnion, nullTpe)
-        } else {
-          ctx.typeComparer.lub(tp1.widenUnion, tp2.widenUnion, canConstrain = true) match {
-            case union: OrType => union.join
-            case res => res
+    def widenUnion(implicit ctx: Context): Type = {
+      def normalizeRightOrNull(tp: OrType, lhs: Type, rhs: Type): Type = lhs match {
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: I think this method can be simplified to

```Scala
// Given a type `tpe`, if it is already a nullable union, return it unchanged.
// Otherwise, construct a nullable union where `tpe` is the lhs (use `orig` to potentially avoid creating a new object for the union). 
def ensureNullableUnion(tpe: Type, orig: OrType): Type = tpe match {
  // Don't destruct `orTpe` to avoid the method call to `unapply`.
  case orTpe: OrType if orTpe.tp2.isNullType => tpe
  case _ => orig.derivedOrType(tpe, defn.NullType)
}
```

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 16]
```diff
@@ -1068,26 +1068,50 @@ object Types {
...
+
+      widen match {
+        case tp @ OrType(lhs, rhs) =>
+          def defaultJoin(tp1: Type, tp2: Type) =
+            ctx.typeComparer.lub(tp1, tp2, canConstrain = true) match {
+              case union: OrType => union.join
+              case res => res
+            }
+          if (ctx.explicitNulls) {
+            // Don't widen `T|Null`, since otherwise we wouldn't be able to infer nullable unions.
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/Types.scala:1095]: As it stands, this is hard to follow because there are so many cases.

I'd propose the following variant:

```Scala
// Test for nullable union that assumes the type has already been normalized.
def isNullableUnionFast(tp: Type): Boolean = tp match {
  // Don't destruct `orTpe` to avoid the method call to `unapply`.
  case orTpe: OrType if orTpe.tp2.isNullType => true
  case _ => false
}

if (rhs.isNullType) ensureNullableUnion(lhs.widenUnion, tp)
else if (lhs.isNullType) ensureNullableUnion(rhs.widenUnion, tp)
else {
  val lhsWiden = lhs.widenUnion
  val rhsWiden = rhs.widenUnion
  val tmpRes = ctx.typeComparer.lub(lhsWiden, rhsWiden, canConstrain = true) match {
    case union: OrType => union.join
    case res => res
  }
  if (isNullableUnionFast(lhsWiden) || isNullableUnionFast(rhsWiden)) ensureNullableUnion(tmpRes)
  else tmpRes
}
```

**Review by abeln — CHANGES_REQUESTED** [review]

File: compiler/src/dotty/tools/dotc/core/Types.scala (hunk 15 above)
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/Types.scala]: I don't think we can simplify this function if we want to track Null and JavaNull.

File: compiler/src/dotty/tools/dotc/core/Types.scala (hunk 16 above)
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/Types.scala:1095]: Same problem here, I think we need to track the Null type

**Comment by noti0na1** [comment]
@abeln Done

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 17]
```diff
@@ -1068,26 +1068,52 @@ object Types {
...
+            case orTpe: OrType if orTpe.tp2.isNullType => true
+            case _ => false
+          }
+
+          if (ctx.explicitNulls) {
+            // Don't widen `T|Null`, since otherwise we wouldn't be able to infer nullable unions.
+            if (rhs.isNullType) ensureNullableUnion(lhs.widenUnion, tp)
+            else if (lhs.isNullType) ensureNullableUnion(rhs.widenUnion, tp)
+            else {
+              val lhsWiden = lhs.widenUnion
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/Types.scala:1101]: Could you please add a comment explaining what the postcondition that we assume is?
e.g.

The correctness of the code below relies on the following postcondition for `widenUnion`.
The result of `widenUnion` is either a non-union type, or a union type where the rhs is the `Null` type.

**Review by abeln — APPROVED** [review]
LGTM, thanks! After adding the comment I mentioned, please feel free to merge.
